src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,420 @@
+};
+static const char *const ridableNames[] = {
+    X_FOREACH_RoomRidableEnum(DECL) //
+};
+static const char *const sundeathNames[] = {
+    X_FOREACH_RoomSundeathEnum(DECL) //
+};
+static const char *const terrainNames[] = {
+    X_FOREACH_RoomTerrainEnum(DECL) //
+};
+#undef DECL
+
+void XmlMapStorage::saveDoorFlags(QXmlStreamWriter &stream, DoorFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    QString list;
+    const char *separator = "";
+    for (uint x = 0; x < sizeof(doorFlagNames) / sizeof(doorFlagNames[0]); x++) {
+        if (fl.contains(DoorFlagEnum(x))) {
+            list += separator;
+            separator = ",";
+            list += doorFlagNames[x];
+        }
+    }
+    saveXmlAttribute(stream, "doorflags", list);
+}
+
+void XmlMapStorage::saveExitFlags(QXmlStreamWriter &stream, ExitFlags fl)
+{
+    fl.remove(ExitFlagEnum::EXIT); // always set, do not save it
+    if (fl.isEmpty()) {
+        return;
+    }
+    QString list;
+    const char *separator = "";
+    for (uint x = 0; x < sizeof(exitFlagNames) / sizeof(exitFlagNames[0]); x++) {
+        if (fl.contains(ExitFlagEnum(x))) {
+            list += separator;
+            separator = ",";
+            list += exitFlagNames[x];
+        }
+    }
+    saveXmlAttribute(stream, "flags", list);
+}
+
+void XmlMapStorage::saveRoomLoadFlags(QXmlStreamWriter &stream, RoomLoadFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    stream.writeStartElement("flags");
+    QString separator;
+    for (uint x = 0; x < sizeof(loadFlagNames) / sizeof(loadFlagNames[0]); x++) {
+        if (fl.contains(RoomLoadFlagEnum(x))) {
+            stream.writeCharacters(separator);
+            if (separator.isEmpty()) {
+                separator = ",";
+            }
+            stream.writeCharacters(loadFlagNames[x]);
+        }
+    }
+    stream.writeEndElement();
+}
+
+void XmlMapStorage::saveRoomMobFlags(QXmlStreamWriter &stream, RoomMobFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    stream.writeStartElement("mobflags");
+    QString separator;
+    for (uint x = 0; x < sizeof(mobFlagNames) / sizeof(mobFlagNames[0]); x++) {
+        if (fl.contains(RoomMobFlagEnum(x))) {
+            stream.writeCharacters(separator);
+            if (separator.isEmpty()) {
+                separator = ",";
+            }
+            stream.writeCharacters(mobFlagNames[x]);
+        }
+    }
+    stream.writeEndElement();
+}
+
+NODISCARD QString XmlMapStorage::alignName(const RoomAlignEnum x)
+{
+    return enumName(uint(x), alignNames, sizeof(alignNames) / sizeof(alignNames[0]));
+}
+
+NODISCARD QString XmlMapStorage::lightName(const RoomLightEnum x)
+{
+    return enumName(uint(x), lightNames, sizeof(lightNames) / sizeof(lightNames[0]));
+}
+
+NODISCARD QString XmlMapStorage::markClassName(const InfoMarkClassEnum e)
+{
+    const uint x = uint(e);
+    const uint count = sizeof(markClassNames) / sizeof(markClassNames[0]);
```

Comment:
Why would there be more mark classes than what the code is aware of? This could mean map corruption!

Reply:
My habit at defensive programming.

Would you prefer an assertion, a throw, or something else?

Reply:
assert() would be an acceptable solution. We shouldn't persist the unknown value to the XML file though. This is where a better schema validator (like a DTD) will help us.

I think we would need to be more paranoid during data loading. We would most likely want to abort at that point and inform the user they corrupted the data.

Reply:
I added `assert(false)` at the end of every `switch (e)` to catch "impossible" enum values.